File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque

def solution(maps):
    s = []
    l = []
    e = []
    m = len(maps)
    n = len(maps[0])
    
    for i in range(m) :
        for j in range(n) :
            if maps[i][j] == "S" : s = [i,j]
            if maps[i][j] == "E" : e = [i,j]
            if maps[i][j] == "L" : l = [i,j]
            
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]

    def bfs(start, end) :
        visited = [[False] * n for _ in range(m)] 
        visited[start[0]][start[1]] = True
        queue = deque()
        queue.append([start[0], start[1], 0])
        
        while queue :
            x, y, c = queue.popleft()
            
            if x == end[0] and y == end[1] :
                return c
            
            for i in range(4) :
                nx = x + dx[i]
                ny = y + dy[i]

                if 0 <= nx < m and 0 <= ny < n:
                    if visited[nx][ny] == False and maps[nx][ny] != "X":
                        visited[nx][ny] = True
                        queue.append([nx,ny,c + 1])
        return -1
    
    dist1 = bfs(s,l)
    dist2 = bfs(l,e)
    
    if dist1 == -1 or dist2 == -1:
        return -1
    
    return dist1 + dist2
        

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    int[][] d = {{0,1}, {0,-1}, {1,0}, {-1,0}};
    int n = 0;
    int m = 0;
    public int solution(String[] maps) {
        n = maps.length;
        m = maps[0].length();
        // S-> L로 가는 최적의 경로
        // L-> E로 가는 최적의 경로
        int[] S = new int[2];
        int[] L = new int[2];
        int[] E = new int[2];
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(maps[i].charAt(j) == 'S') S = new int[]{i, j};
                if(maps[i].charAt(j) == 'L') L = new int[]{i, j};
                if(maps[i].charAt(j) == 'E') E = new int[]{i, j};
            }
        }
        
        int dist1 = bfs(S, L, maps, new boolean[n][m]);
        int dist2 = bfs(L, E, maps, new boolean[n][m]);
        
        if(dist1 == 0 || dist2 == 0) return -1;
        return dist1 + dist2;
    }
    
    public int bfs(int[] start, int[] end, String[] maps, boolean[][] v){
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{start[0], start[1], 0});
        v[start[0]][start[1]] = true;
        
        while(!q.isEmpty()){
            int[] now = q.poll();
            
            if(now[0] == end[0] && now[1] == end[1]) return now[2];
            
            for(int i = 0; i < 4; i++){
                int nx = now[0] + d[i][0];
                int ny = now[1] + d[i][1];
                
                if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                
                if(!v[nx][ny] && maps[nx].charAt(ny) != 'X'){
                    q.offer(new int[]{nx, ny, now[2] + 1});
                    v[nx][ny] = true;
                }
            }
        }
        return 0;
    }
}